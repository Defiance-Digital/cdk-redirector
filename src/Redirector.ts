import { EndpointType, MockIntegration, PassthroughBehavior, RestApi } from 'aws-cdk-lib/aws-apigateway';
import { ICertificate } from 'aws-cdk-lib/aws-certificatemanager';
import { Construct } from 'constructs';

export interface RedirectorProps {
  /**
   * A certificate to associate with the API Gateway Endpoint
   */
  readonly certificate: ICertificate;

  /**
   * The domain name to associate with the API Gateway Endpoint. This domain needs to be covered by the given certificate.
   */
  readonly domainName: string;

  /**
   * The type of redirect to perform. Either a permanent (301) or temporary (302) redirect.
   *
   * @default REDIRECT_TYPE.TEMPORARY
   */
  readonly redirectType?: RedirectType;


  /**
   * The URL to redirect to.
   */
  readonly redirectUrl: string;

  /**
   * The name of the API Gateway Endpoint. If not given, then 'Redirecter' is used.
   *
   * @optional
   */
  readonly restApiName?: string;
}

export enum RedirectType {
  TEMPORARY = 'temporary',
  PERMANENT = 'permanent',
}

/**
 * Creates an API Gateway endpoint that redirects to a given URL.
 */
export class Redirector extends Construct {
  constructor(scope: Construct, id: string, props: RedirectorProps) {
    super(scope, id);
    const restApi = new RestApi(this, 'Resource', {
      domainName: {
        domainName: props.domainName,
        certificate: props.certificate,
      },
      deployOptions: {
        stageName: 'v1',
      },
      restApiName: props.restApiName ?? 'Redirector',
      endpointTypes: [EndpointType.REGIONAL],
    });

    const statusCode = props.redirectType === RedirectType.PERMANENT ? 301 : 302;

    restApi.root.addMethod('GET', new MockIntegration({
      integrationResponses: [{
        statusCode: `${statusCode}`,
        responseParameters: {
          'method.response.header.Location': `'${props.redirectUrl}'`,
        },
      }],
      passthroughBehavior: PassthroughBehavior.NEVER,
      requestTemplates: {
        'application/json': `{"statusCode": ${statusCode}}`,
      },
    }), {
      methodResponses: [
        {
          statusCode: `${statusCode}`,
          responseParameters: {
            'method.response.header.Location': true,
          },
        },
      ],
    });
  }

}
