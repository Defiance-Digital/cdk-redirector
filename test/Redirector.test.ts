import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Certificate } from 'aws-cdk-lib/aws-certificatemanager';
import { RedirectType, Redirector, RedirectorProps } from '../src';

describe('Redirector', () => {
  function createTestStack(props?: Partial<RedirectorProps>) {
    const app = new App();
    const stack = new Stack(app, 'Stack');
    new Redirector(stack, 'Redirector', {
      certificate: new Certificate(stack, 'Certificate', {
        domainName: 'someolddomain.com',
      }),
      domainName: 'someolddomain.com',
      redirectUrl: 'somenewdomain.com',
      ...props,
    });

    const template = Template.fromStack(stack);
    return template;
  }

  describe('snapshots', () => {
    it('with default props', () => {
      const template = createTestStack();
      expect(template.toJSON()).toMatchSnapshot();
    });
  });
  describe('restApiName', () => {
    it('Uses default name with restApiName not provided', () => {
      const template = createTestStack();
      template.hasResourceProperties('AWS::ApiGateway::RestApi', {
        EndpointConfiguration: {
          Types: [
            'REGIONAL',
          ],
        },
        Name: 'Redirector',
      });
    });

    it('Uses given name with restApiName provided', () => {
      const template = createTestStack({
        restApiName: 'MyRedirector',
      });
      template.hasResourceProperties('AWS::ApiGateway::RestApi', {
        EndpointConfiguration: {
          Types: [
            'REGIONAL',
          ],
        },
        Name: 'MyRedirector',
      });
    });

  });

  it('Custom Domain is correct', () => {
    const template = createTestStack();
    template.hasResourceProperties('AWS::ApiGateway::DomainName', {
      DomainName: 'someolddomain.com',
      EndpointConfiguration: {
        Types: [
          'REGIONAL',
        ],
      },
    });
  });
  describe('GET Method', () => {
    it('with 302 temp redirect', () => {
      const template = createTestStack();
      template.hasResourceProperties('AWS::ApiGateway::Method', {
        AuthorizationType: 'NONE',
        HttpMethod: 'GET',
        Integration: {
          IntegrationResponses: [
            {
              ResponseParameters: {
                'method.response.header.Location': "'somenewdomain.com'",
              },
              StatusCode: '302',
            },
          ],
          PassthroughBehavior: 'NEVER',
          RequestTemplates: {
            'application/json': '{"statusCode": 302}',
          },
          Type: 'MOCK',
        },
        MethodResponses: [
          {
            ResponseParameters: {
              'method.response.header.Location': true,
            },
            StatusCode: '302',
          },
        ],

      });
    });
    it('with 301 perm redirect', () => {
      const template = createTestStack({
        redirectType: RedirectType.PERMANENT,
      });
      template.hasResourceProperties('AWS::ApiGateway::Method', {
        AuthorizationType: 'NONE',
        HttpMethod: 'GET',
        Integration: {
          IntegrationResponses: [
            {
              ResponseParameters: {
                'method.response.header.Location': "'somenewdomain.com'",
              },
              StatusCode: '301',
            },
          ],
          PassthroughBehavior: 'NEVER',
          RequestTemplates: {
            'application/json': '{"statusCode": 301}',
          },
          Type: 'MOCK',
        },
        MethodResponses: [
          {
            ResponseParameters: {
              'method.response.header.Location': true,
            },
            StatusCode: '301',
          },
        ],
      });
    });
  });

});
