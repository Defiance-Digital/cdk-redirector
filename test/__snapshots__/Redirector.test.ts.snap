// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Redirector snapshots with default props 1`] = `
{
  "Outputs": {
    "RedirectorEndpoint39B84C17": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "RedirectorA15E92ED",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "RedirectorDeploymentStagev12544658A",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Certificate4E7ABB08": {
      "Properties": {
        "DomainName": "someolddomain.com",
        "DomainValidationOptions": [
          {
            "DomainName": "someolddomain.com",
            "ValidationDomain": "someolddomain.com",
          },
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "Stack/Certificate",
          },
        ],
        "ValidationMethod": "EMAIL",
      },
      "Type": "AWS::CertificateManager::Certificate",
    },
    "RedirectorA15E92ED": {
      "Properties": {
        "EndpointConfiguration": {
          "Types": [
            "REGIONAL",
          ],
        },
        "Name": "Redirector",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "RedirectorAccountB92FA666": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "RedirectorA15E92ED",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "RedirectorCloudWatchRole825AFF89",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "RedirectorCloudWatchRole825AFF89": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "RedirectorCustomDomainC3D1481F": {
      "Properties": {
        "DomainName": "someolddomain.com",
        "EndpointConfiguration": {
          "Types": [
            "REGIONAL",
          ],
        },
        "RegionalCertificateArn": {
          "Ref": "Certificate4E7ABB08",
        },
      },
      "Type": "AWS::ApiGateway::DomainName",
    },
    "RedirectorCustomDomainMapStackRedirector14D4840B661510D0": {
      "Properties": {
        "DomainName": {
          "Ref": "RedirectorCustomDomainC3D1481F",
        },
        "RestApiId": {
          "Ref": "RedirectorA15E92ED",
        },
        "Stage": {
          "Ref": "RedirectorDeploymentStagev12544658A",
        },
      },
      "Type": "AWS::ApiGateway::BasePathMapping",
    },
    "RedirectorDeployment20B89C48756f86bc017503e7c6f29ea7ee62353b": {
      "DependsOn": [
        "RedirectorGET003FDA46",
      ],
      "Properties": {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": {
          "Ref": "RedirectorA15E92ED",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "RedirectorDeploymentStagev12544658A": {
      "DependsOn": [
        "RedirectorAccountB92FA666",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "RedirectorDeployment20B89C48756f86bc017503e7c6f29ea7ee62353b",
        },
        "RestApiId": {
          "Ref": "RedirectorA15E92ED",
        },
        "StageName": "v1",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "RedirectorGET003FDA46": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": {
          "IntegrationResponses": [
            {
              "ResponseParameters": {
                "method.response.header.Location": "'somenewdomain.com'",
              },
              "StatusCode": "302",
            },
          ],
          "PassthroughBehavior": "NEVER",
          "RequestTemplates": {
            "application/json": "{"statusCode": 302}",
          },
          "Type": "MOCK",
        },
        "MethodResponses": [
          {
            "ResponseParameters": {
              "method.response.header.Location": true,
            },
            "StatusCode": "302",
          },
        ],
        "ResourceId": {
          "Fn::GetAtt": [
            "RedirectorA15E92ED",
            "RootResourceId",
          ],
        },
        "RestApiId": {
          "Ref": "RedirectorA15E92ED",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
